File: backend/services/provider_manager.py
"""
==========================================================
Easy AI Studio
File    : backend/services/provider_manager.py
Version : 1.0.0
Sprint  : 10
==========================================================
"""

from typing import Dict
from typing import List
from typing import Optional


class ProviderManager:

    def __init__(self):

        self.providers: Dict[str, dict] = {}

    def register(

        self,

        provider_id: str,

        provider: dict

    ) -> None:

        self.providers[provider_id] = provider

    def unregister(

        self,

        provider_id: str

    ) -> bool:

        if provider_id in self.providers:

            del self.providers[provider_id]

            return True

        return False

    def get(

        self,

        provider_id: str

    ) -> Optional[dict]:

        return self.providers.get(provider_id)

    def exists(

        self,

        provider_id: str

    ) -> bool:

        return provider_id in self.providers

    def all(self) -> List[dict]:

        return list(self.providers.values())

    def ids(self) -> List[str]:

        return list(self.providers.keys())

    def count(self) -> int:

        return len(self.providers)

    def clear(self):

        self.providers.clear()

    def default(self):

        if "openai" in self.providers:

            return self.providers["openai"]

        if len(self.providers) > 0:

            return next(

                iter(

                    self.providers.values()

                )

            )

        return None


provider_manager = ProviderManager()

File: frontend_old/js/providerManager.js
/* ==========================================================
   Easy AI Studio
   File: frontend/js/providerManager.js
   Version: 1.0.0
   Sprint: 6 - AI Provider Manager
   ========================================================== */

"use strict";

/* ==========================================================
   Provider Manager
   ========================================================== */

const ProviderManager = {

    providers: new Map(),

    activeProvider: null,

    /* ======================================================
       Initialize
       ====================================================== */

    init() {

        this.registerDefaults();

        console.log(

            "Provider Manager Ready"

        );

    },

    /* ======================================================
       Register Default Providers
       ====================================================== */

    registerDefaults() {

        this.register({

            id: "ollama",

            name: "Ollama",

            manager: window.OllamaManager

        });

        this.register({

            id: "openai",

            name: "OpenAI",

            manager: window.OpenAIManager

        });

        this.register({

            id: "gemini",

            name: "Gemini",

            manager: window.GeminiManager

        });

        this.register({

            id: "claude",

            name: "Claude",

            manager: window.ClaudeManager

        });

        this.register({

            id: "grok",

            name: "Grok",

            manager: window.GrokManager

        });

        this.register({

            id: "stable-diffusion",

            name: "Stable Diffusion",

            manager: window.StableDiffusionManager

        });

        this.register({

            id: "comfyui",

            name: "ComfyUI",

            manager: window.ComfyUIManager

        });

        this.register({

            id: "flux",

            name: "FLUX",

            manager: window.FluxManager

        });

    },

    /* ======================================================
       Register Provider
       ====================================================== */

    register(provider) {

        this.providers.set(

            provider.id,

            provider

        );

    },

    /* ======================================================
       Active Provider
       ====================================================== */

    setActive(id) {

        if (

            !this.providers.has(id)

        ) {

            return;

        }

        this.activeProvider = id;

        EventBus.emit(

            "provider.changed",

            this.get(id)

        );

    },

    /* ======================================================
       Current Provider
       ====================================================== */

    current() {

        return this.get(

            this.activeProvider

        );

    },

    /* ======================================================
       Find Provider
       ====================================================== */

    get(id) {

        return this.providers.get(id);

    },

    /* ======================================================
       List Providers
       ====================================================== */

    all() {

        return Array.from(

            this.providers.values()

        );

    }

};

/* ==========================================================
   Global
   ========================================================== */

window.ProviderManager = ProviderManager;